cpp/src/arrow/flight/transport_server_impl.cc

```diff
@@ -0,0 +1,320 @@
+    ++stats_.num_record_batches;
+    return Status::OK();
+  }
+
+  Status Close() override {
+    // It's fine to Close() without writing data
+    return Status::OK();
+  }
+
+  ipc::WriteStats stats() const override { return stats_; }
+
+ private:
+  Status WritePayload(const FlightPayload& payload) {
+    RETURN_NOT_OK(stream_->WriteData(payload));
+    ++stats_.num_messages;
+    return Status::OK();
+  }
+
+  Status CheckStarted() {
+    if (!started_) {
+      return Status::Invalid("This writer is not started. Call Begin() with a schema");
+    }
+    return Status::OK();
+  }
+
+  Status EnsureDictionariesWritten(const RecordBatch& batch) {
+    if (dictionaries_written_) {
+      return Status::OK();
+    }
+    dictionaries_written_ = true;
+    ARROW_ASSIGN_OR_RAISE(const auto dictionaries,
+                          ipc::CollectDictionaries(batch, mapper_));
+    for (const auto& pair : dictionaries) {
+      FlightPayload payload{};
+      RETURN_NOT_OK(ipc::GetDictionaryPayload(pair.first, pair.second, ipc_options_,
+                                              &payload.ipc_message));
+      RETURN_NOT_OK(WritePayload(payload));
+      ++stats_.num_dictionary_batches;
+    }
+    return Status::OK();
+  }
+
+  ServerDataStream* stream_;
+  ::arrow::ipc::IpcWriteOptions ipc_options_;
+  ipc::DictionaryFieldMapper mapper_;
+  ipc::WriteStats stats_;
+  bool started_ = false;
+  bool dictionaries_written_ = false;
+};
+
+/// \brief Adapt TransportDataStream to the FlightMetadataWriter
+///   interface for DoPut.
+class TransportMetadataWriter final : public FlightMetadataWriter {
+ public:
+  explicit TransportMetadataWriter(ServerDataStream* stream) : stream_(stream) {}
+
+  Status WriteMetadata(const Buffer& buffer) override {
+    return stream_->WritePutMetadata(buffer);
+  }
+
+ private:
+  ServerDataStream* stream_;
+};
+}  // namespace
+
+Status ServerTransportImpl::DoGet(const ServerCallContext& context, const Ticket& ticket,
+                                  ServerDataStream* stream) {
+  std::unique_ptr<FlightDataStream> data_stream;
+  RETURN_NOT_OK(base_->DoGet(context, ticket, &data_stream));
+
+  if (!data_stream) return Status::KeyError("No data in this flight");
+
+  // Write the schema as the first message in the stream
+  FlightPayload schema_payload;
+  RETURN_NOT_OK(data_stream->GetSchemaPayload(&schema_payload));
+  auto status = stream->WriteData(schema_payload);
+  // Connection terminated
+  if (status.IsIOError()) return Status::OK();
+  RETURN_NOT_OK(status);
+
+  // Consume data stream and write out payloads
+  while (true) {
+    FlightPayload payload;
+    RETURN_NOT_OK(data_stream->Next(&payload));
+    // End of stream
+    if (payload.ipc_message.metadata == nullptr) break;
+    auto status = stream->WriteData(payload);
+    // Ignore IOError (used to signal that client disconnected; there's nothing
+    // we can do - e.g. see WritePayload in serialization_internal.cc)
```

Comment:
IOError can also indicate decompression errors and other issues. Can we simply let the caller decide what to do with the error instead of silencing it here? (same for the other test a few lines above)

Reply:
Sounds good. Actually, I'll change the API to return Result<bool> or something so we can differentiate between errors and end-of-stream.

Reply:
Done, thanks for reminding me to take the chance to clean up some of these things.